cjdns-bytes/src/utils.rs

```diff
@@ -1,150 +1,224 @@
-        Ok(u64::from_be_bytes(bytes_array))
-    }
+        pub fn read_u16_be(&mut self) -> Result<u16> {
+            let bytes = self.read_slice(2)?;
+            let bytes_array = <[u8; 2]>::try_from(bytes).expect("invalid slice size");
+            Ok(u16::from_be_bytes(bytes_array))
+        }
 
-    pub fn read_array_32(&mut self) -> Result<[u8; 32]> {
-        let bytes_32_slice = self.take_bytes(32)?;
-        let bytes_32_array = <[u8; 32]>::try_from(bytes_32_slice).expect("invalid slice size");
-        Ok(bytes_32_array)
-    }
+        pub fn read_u32_be(&mut self) -> Result<u32> {
+            let bytes = self.read_slice(4)?;
+            let bytes_array = <[u8; 4]>::try_from(bytes).expect("invalid slice size");
+            Ok(u32::from_be_bytes(bytes_array))
+        }
 
-    pub fn read_all_pure(&self) -> &[u8] {
-        self.0
-    }
+        pub fn read_u64_be(&mut self) -> Result<u64> {
+            let bytes = self.read_slice(8)?;
+            let bytes_array = <[u8; 8]>::try_from(bytes).expect("invalid slice size");
+            Ok(u64::from_be_bytes(bytes_array))
+        }
 
-    pub fn read_all_mut(&mut self) -> &[u8] {
-        self.take_bytes(self.len()).expect("attempting to read data more than slice have")
-    }
+        pub fn read_array_32(&mut self) -> Result<[u8; 32]> {
+            let bytes_32_slice = self.read_slice(32)?;
+            let bytes_32_array = <[u8; 32]>::try_from(bytes_32_slice).expect("invalid slice size");
+            Ok(bytes_32_array)
+        }
 
-    pub fn take_bytes(&mut self, count: usize) -> Result<&[u8]> {
-        if self.len() < count {
-            return Err(());
+        pub fn pick_remainder(&self) -> &'a [u8] {
+            self.0
+        }
+
+        pub fn read_remainder(&mut self) -> &'a [u8] {
+            self.read_slice(self.len()).expect("attempting to read data more than slice have")
+        }
+
+        pub fn read_slice(&mut self, count: usize) -> Result<&'a [u8]> {
+            if self.len() < count {
+                return Err(InsufficientBuffer);
+            }
+            let (ret_bytes, rest_bytes) = self.0.split_at(count);
+            self.0 = rest_bytes;
+            Ok(ret_bytes)
+        }
+
+        fn len(&self) -> usize {
+            self.0.len()
         }
-        let (ret_bytes, rest_bytes) = self.0.split_at(count);
-        self.0 = rest_bytes;
-        Ok(ret_bytes)
     }
 
-    fn len(&self) -> usize {
-        self.0.len()
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+
+        #[test]
+        fn test_reader() {
+            let bytes = vec![0u8; 64];
+            let mut reader = Reader::new(&bytes);
+
+            let array_32 = reader.read_array_32().expect("bytes are read out");
+            assert_eq!(array_32, [0; 32]);
+            let num64 = reader.read_u64_be().expect("bytes are read out");
+            assert_eq!(num64.to_be_bytes(), [0; 8]);
```

Comment:
So here you compare with [0x01, 0x02, 0x03, 0x04] which is much more reliable than four zeroes.
etc.

Reply:
done